File: deli.java
//Author: Atharva Nagarkar
//Solution to David and the Deli

import java.io.*;
import java.util.*;

public class deli {
	
	final int log = 31;
	
	public void solve(FS in, PrintWriter out) {
		//read in number of cases
		int cases = in.nextInt();
		
		//process each case
		for(int t = 1; t <= cases; ++t) {
			//header for each case
			out.printf("Scenario #%d:%n", t);
			
			//read in number of elements and number of queries
			int n = in.nextInt();
			int m = in.nextInt();
			
			//make an array for the cost of each item
			long[] a = new long[n];
			
			//keep 2 arrays of Binary Indexed Trees
			//each Tree will keep an answer for it's respective
			//bit and and all bits under it
			
			//sums[i].sum(L, R) will keep the sum of costs from 
			//index L to R of items whose cost is less than or 
			//equal to 2^i
			BIT[] sums = new BIT[log];
			
			//cnts[i].sum(L, R) will keep the number of items
			//from index L to R whose cost is less than or equal
			//to 2^i
			BIT[] cnts = new BIT[log];
			
			//initialize our binary indexed trees
			for(int i = 0; i < log; ++i) {
				sums[i] = new BIT(n);
				cnts[i] = new BIT(n);
			}
			
			//read in the costs for the items
			for(int i = 0; i < n; ++i) {
				a[i] = in.nextLong(); //cost for item i
				//update the corresponding trees
				for(int j = 0; j < log; ++j) {
					long pow = 1L << j;
					if(a[i] > pow) continue; //if 2^j is too small, then continue
					sums[j].update(i, a[i]); //add the value of a[i] to the i'th index of the j'th tree
					cnts[j].update(i, 1); //add 1 to the i'th index of the j'th tree
				}
			}
			
			//let's process queries
			while(m-->0) {
				int type = in.nextInt(); //what type of query
				if(type == 1) { //give max number of items that can be chosen
					long budget = in.nextLong(); //see what his budget is for the query
					int idx = 0; //keep track of the item he is currently viewing to buy, but has not bought yet
					int ans = 0; //out answer to the query, the number of things he can buy
					for(int bit = log - 1; idx < n && bit >= 0; --bit) { //loop through all bits in the binary representation of the budget
						long pow = 1L << bit;
						if(pow > budget) continue; //if this amount is too big for the budget, we can't buy anything with this cost
						int upto = sums[bit].lift(budget, idx); //see how far right we can go given our current budget, starting at idx (our current index)
						ans += cnts[bit].sum(idx, upto - 1); //add the number of valid items in this range for the current bit to our answer
						budget -= sums[bit].sum(idx, upto - 1); //take out the sum of costs for taking the items
						idx = upto; //update current index to the new thing we are looking at
					}
					out.printf("David can get %d toppings%n", ans); //output the answer
				} 
				else { //update query
					int ind = in.nextInt() - 1; //see which index we are updating
					long x = in.nextLong(); //our new value
					long prev = a[ind]; //our current value
					
					//first lets take out all the contributions of the previous value
					for(int bit = 0; bit < log; ++bit) {
						long pow = 1L << bit;
						if(pow < prev) continue;
						sums[bit].update(ind, -prev); //update the sum contribution
						cnts[bit].update(ind, -1); //also the count contribution
					}
					
					//now lets update all things that the new value will contribute to
					for(int bit = 0; bit < log; ++bit) {
						long pow = 1L << bit;
						if(pow < x) continue;
						sums[bit].update(ind, x); //update the sum contribution
						cnts[bit].update(ind, 1); //also the count contribution
					}
					
					a[ind] = x; //update our value in the array to the new one
				}
			}
			out.println(); //blank line after each scenario
		}
	}
	
	public static void main(String[] args) {
		FS in = new FS();
		PrintWriter out = new PrintWriter(System.out);
		new deli().solve(in, out);
		out.close();
	}
	
	//a standard binary indexed tree for log(n) range sum and log(n) updates
	class BIT {
		int n;
		long[] tree;
		
		public BIT(int nn) {
			n = nn + 10;
			tree = new long[n + 1];
		}
		
		//binary search the tree for the first index i such that the sum from
		//l to i is strictly greater than v
		int lift(long v, int l) {
			v += sum(0, l - 1);
			long sum = 0;
			int pos = 0;
			for(int i = log; i >= 0; --i) {
				if(pos + (1L << i) >= n) continue;
				if(sum + tree[pos + (1 << i)] > v) continue;
				sum += tree[pos + (1 << i)];
				pos += (1 << i);
			}
			return 1 + pos - 5;
		}
		
		long read(int i) {
			i += 5;
			long sum = 0;
			while (i > 0) {
				sum += tree[i];
				i -= i & -i;
			}
			return sum;
		}
		
		long sum(int l, int r) {
			if(r < l) return 0;
			return read(r) - read(l - 1);
		}
		
		void update(int i, long val) {
			i += 5;
			while (i <= n) {
				tree[i] += val;
				i += i & -i;
			}
		}
	}
	
	static class FS {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer("");
		String next() {
			try {
				while(!st.hasMoreElements()) {
					st = new StringTokenizer(br.readLine());
				}
			} catch(Exception e) {}
			return st.nextToken();
		}
		int nextInt() {
			return Integer.parseInt(next());
		}
		long nextLong() {
			return Long.parseLong(next());
		}
	}
}

File: deli.py
# Python implementation of Binary Indexed Tree - Courtesy of Geeksforgeeks

# https://www.geeksforgeeks.org/binary-indexed-tree-or-fenwick-tree-2/
  
# Returns sum of arr[0..index]. This function assumes 
# that the array is preprocessed and partial sums of 
# array elements are stored in BITree[]. 

log = 31

# This is unique to this problem
# Binary search the tree for the first index i such that the sum from 0 to i-1 is strictly greater than v
def lift(BITTree, n, i, v):
	v += sum(BITTree, 0, i-1)
	tot = 0
	pos = 0
	for i in reversed(range(log+1)):
		if(pos + (1 << i) > n):
			continue
		if(tot + BITTree[pos + (1 << i)] > v):
			continue
		tot += BITTree[pos+(1 << i)]
		pos += (1 << i)
	return pos;

def sum(BITTree, l, r):
	if(r < l):
		return 0
	return getsum(BITTree, r)-getsum(BITTree, l-1)

def getsum(BITTree,i): 
    s = 0 #initialize result 
  
    # index in BITree[] is 1 more than the index in arr[] 
    i = i+1
  
    # Traverse ancestors of BITree[index] 
    while i > 0: 
  
        # Add current element of BITree to sum 
        s += BITTree[i] 
  
        # Move index to parent node in getSum View 
        i -= i & (-i) 
    return s 
  
# Updates a node in Binary Index Tree (BITree) at given index 
# in BITree. The given value 'val' is added to BITree[i] and 
# all of its ancestors in tree. 
def updatebit(BITTree , n , i ,v): 
  
    # index in BITree[] is 1 more than the index in arr[] 
    i += 1
  
    # Traverse all ancestors and add 'val' 
    while i <= n: 
  
        # Add 'val' to current node of BI Tree 
        BITTree[i] += v 
  
        # Update index to that of parent in update View 
        i += i & (-i)
  
  
# Constructs and returns a Binary Indexed Tree for given 
# array of size n. 
def construct(arr, n): 
  
    # Create and initialize BITree[] as 0 
    BITTree = [0]*(n+1) 
  
    # Store the actual values in BITree[] using update() 
    for i in range(n): 
        updatebit(BITTree, n, i, arr[i]) 
  
    return BITTree 


cases = int(input())
for q in range(cases):
	print('Scenario #'+str(q+1)+':')
	inp = input().split(" ")
	n = int(inp[0])
	m = int(inp[1])
	empty = [0]*n
	a = [0]*n
	
	sums = [construct(empty, n) for i in range(log)]
	cnts = [construct(empty, n) for i in range(log)]
	
	inp = input().split(" ")
	for i in range(n):
		a[i] = int(inp[i])
		
		for j in range(log):
			pow = 1 << j
			if(a[i] > pow):
				continue
			updatebit(sums[j], n, i, a[i])
			updatebit(cnts[j], n, i, 1)
			
	for query in range(m):
		inp = input().split(" ")
		type = int(inp[0])
		if(type == 1):
			budget = int(inp[1])
			idx = 0
			ans = 0
			bit = 30
			while(bit >= 0 and idx < n and budget > 0):
				pow = (1 << bit) 
				if(pow > budget):
					bit = bit-1
					continue
				upto = lift(sums[bit], n, idx, budget)
				ans += sum(cnts[bit], idx, upto-1)
				budget -= sum(sums[bit], idx, upto-1)
				idx = upto
				bit = bit - 1
			print('David can get '+str(ans)+' toppings')
		else:
			idx = int(inp[1])-1
			x = int(inp[2])
			prev = a[idx]
			
			for i in range(log):
				pow = (1 << i)
				if(pow < prev):
					continue
				updatebit(sums[i], n, idx, -prev)
				updatebit(cnts[i], n, idx, -1)
			
			for i in range(log):
				pow = (1 << i)
				if(pow < x):
					continue
				updatebit(sums[i], n, idx, x)
				updatebit(cnts[i], n, idx, 1)
			
			a[idx] = x
	print()
